src/sentry/static/sentry/app/views/organizationContext.tsx

```diff
@@ -51,9 +51,107 @@ type State = {
   errorType?: string | null;
   error?: RequestError | null;
   hooks?: React.ReactNode[];
+  prevProps: {
+    orgId: string;
+    organizationsLoading: boolean;
+    location: RouteComponentProps<{orgId: string}, {}>['location'];
+  };
 };
 
 class OrganizationContext extends React.Component<Props, State> {
+  static getDerivedStateFromProps(props: Props, prevState: State): State {
+    const {prevProps} = prevState;
+
+    if (OrganizationContext.shouldRemount(prevProps, props)) {
+      return OrganizationContext.getDefaultState(props);
+    }
+
+    return {
+      ...prevState,
+      prevProps: {
+        orgId: props.params.orgId,
+        organizationsLoading: props.organizationsLoading,
+        location: props.location,
+      },
+    };
+  }
+
+  static shouldRemount(prevProps: State['prevProps'], props): boolean {
+    const hasOrgIdAndChanged =
+      prevProps.orgId && props.params.orgId && prevProps.orgId !== props.params.orgId;
+
+    const hasOrgId =
+      props.params.orgId ||
+      (props.useLastOrganization && ConfigStore.get('lastOrganization'));
+
+    // protect against the case where we finish fetching org details
+    // and then `OrganizationsStore` finishes loading:
+    // only fetch in the case where we don't have an orgId
+    //
+    // Compare `getOrganizationSlug`  because we may have a last used org from server
+    // if there is no orgId in the URL
+    const organizationLoadingChanged =
+      prevProps.organizationsLoading !== props.organizationsLoading &&
+      props.organizationsLoading === false;
+
+    return (
+      hasOrgIdAndChanged ||
+      (!hasOrgId && organizationLoadingChanged) ||
+      (props.location.state === 'refresh' && prevProps.location.state !== 'refresh')
+    );
+  }
+
+  static getDefaultState(props: Props): State {
+    const prevProps = {
+      orgId: props.params.orgId,
+      organizationsLoading: props.organizationsLoading,
+      location: props.location,
+    };
+
+    if (OrganizationContext.isOrgStorePopulatedCorrectly(props)) {
+      // retrieve initial state from store
+      return {
+        ...OrganizationStore.get(),
+        prevProps,
+      };
+    }
+    return {
+      loading: true,
+      error: null,
+      errorType: null,
+      organization: null,
+      prevProps,
+    };
+  }
+
+  static getOrganizationSlug(props: Props) {
+    return (
+      props.params.orgId ||
+      ((props.useLastOrganization &&
+        (ConfigStore.get('lastOrganization') ||
+          props.organizations?.[0]?.slug)) as string)
+    );
+  }
+
+  static isOrgChanging(props: Props) {
+    const {organization} = OrganizationStore.get();
+    return (
+      organization && organization.slug !== OrganizationContext.getOrganizationSlug(props)
```

Comment:
```suggestion
      organization?.slug !== OrganizationContext.getOrganizationSlug(props)
```

Reply:
This caused the amended test to fail. Instead of being clever, I refactored this to make it explicitly return a boolean.